### PR TITLE
LP-83 first set of changes for make Pausable.  Adding openzepplin pausable.

### PR DESCRIPTION
**Pull request by godfreyhobbs**

### CODE REVIEW

**Comment by libertylocked** [comment]
Hub functions are only callable by owner, so there's no need to pause them.
Maybe add pausable to other contracts, like users (register), records (upload, attest, etc), permissions (share)

I'm not sure if it's necessary to replace expect throw with assert revert. It may be a nice to have thing. Maybe in another PR?

**Comment by godfreyhobbs** [comment]
Will update other contracts in follow-up PR.

File: test/4_permissions_test.js [hunk 1]
```diff
@@ -135,4 +135,22 @@ contract("LinniaPermissions", (accounts) => {
         provider2, { from: patient2 }))
     })
   })
+  describe("pausable", () => {
+    it("should not allow non-admin to pause or unpause", async () => {
+      await expectThrow(instance.pause({ from: accounts[1] }))
+      await expectThrow(instance.unpause({ from: accounts[1] }))
+    })
+    it("should allow admin to pause and unpause", async () => {
```
[the remark this answers is not in the document]
Comment (libertylocked) [reply test/4_permissions_test.js]: maybe *should not allow sharing records when paused*

File: contracts/LinniaPermissions.sol [hunk 2]
```diff
@@ -53,12 +55,13 @@ contract LinniaPermissions is Ownable {
...
         onlyUser
         onlyRecordOwnerOf(dataHash)
+        whenNotPaused
         external
         returns (bool)
     {
         // validate input
         require(viewer != 0);
-        require(keccak256(dataUri) != keccak256(""));
+        require(keccak256(abi.encodePacked(dataUri)) != keccak256(abi.encodePacked("")));
```
Comment (libertylocked) [inline contracts/LinniaPermissions.sol]: what is being packed here?

File: contracts/LinniaRecords.sol [hunk 3]
```diff
@@ -67,6 +68,7 @@ contract LinniaRecords is Ownable {
         bytes32 dataHash, address owner, address attestator,
         string metadata, string dataUri)
         onlyOwner
+        whenNotPaused
```
Comment (libertylocked) [inline contracts/LinniaRecords.sol]: not sure if needed

File: contracts/LinniaRecords.sol [hunk 4]
```diff
@@ -185,7 +191,7 @@ contract LinniaRecords is Ownable {
     {
         // validate input
         require(dataHash != 0);
-        require(keccak256(dataUri) != keccak256(""));
+        require(keccak256(abi.encodePacked(dataUri)) != keccak256(abi.encodePacked("")));
```
Comment (libertylocked) [inline contracts/LinniaRecords.sol]: maybe also check dataUri != "0x"?

File: contracts/LinniaPermissions.sol (hunk 2 above)
Comment (libertylocked) [reply contracts/LinniaPermissions.sol]: convert string to bytes instead of abi encode packed

File: contracts/LinniaRecords.sol (hunk 4 above)
Comment (libertylocked) [reply contracts/LinniaRecords.sol]: I'm not sure why you can't just `require(dataUri != "")`

File: contracts/LinniaRecords.sol [hunk 5]
```diff
@@ -185,7 +191,7 @@ contract LinniaRecords is Ownable {
...
         bytes32 metadataHash = keccak256(abi.encodePacked(metadata));
```
Comment (libertylocked) [inline contracts/LinniaRecords.sol]: bytes(metadata)

Comment (godfreyhobbs) [reply contracts/LinniaRecords.sol]: I am following solidity documentation recommendations by using `abi.encodePacked`

File: contracts/LinniaRecords.sol (hunk 4 above)
Comment (godfreyhobbs) [reply contracts/LinniaRecords.sol]: TypeError: Operator != not compatible with types string memory and literal_string ""
        require(dataUri != "");

 dataUri != "0x" could be a new issue.

File: contracts/LinniaRecords.sol (hunk 5 above)
Comment (libertylocked) [reply contracts/LinniaRecords.sol]: Where in documentation does it say use encode packed for single argument? Should convert to bytes

Comment (godfreyhobbs) [reply contracts/LinniaRecords.sol]: Here is the warning:
```Warning: This function only accepts a single "bytes" argument. Please use "abi.encodePacked(...)" or a similar function to encode the data.```

**Comment by godfreyhobbs** [comment]
I am going to `merge` this `blocked` PR in the morning, as I feel that I have addressed all `requested changes` and this is a pretty simple PR and it is now blocking other work.  

The `abi.encodePacked` is actually part of the already approved PR #29 to 0.4.24.  It only ended up here as the other PR took 11 days to get `approval`

**Comment by libertylocked** [comment]
#29 wasn't a priority and I did not approve it.

Unfortunately I did not see that you are packing a single argument. The correct way is to covert it to bytes

File: contracts/LinniaRecords.sol (hunk 3 above)
Comment (godfreyhobbs) [reply contracts/LinniaRecords.sol]: This also makes it easier for many admins to stay in sync.  So an admin and backup admin have less to coordinate.   
Pausing should result in no big state changes such as adding records.  

The admin will have to unpause if they want to call this method.  Not really a big deal.

**Pull request merged**